File: Leetcode/238/238.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = (int)nums.size();
        vector<int> pre(n);
        pre[0] = nums[0];
        for (int i = 1; i < n; ++i) {
            pre[i] = pre[i - 1] * nums[i];
        }
        vector<int> suff(n);
        suff[0] = nums[n - 1];
        for (int i = 1; i < n; ++i) {
            suff[i] = suff[i - 1] * nums[n - 1 - i];
        }
        nums[0] = suff[n - 2];
        for (int i = 1; i < n - 1; ++i) {
            nums[i] = pre[i - 1] * suff[n - i - 2]; 
        }
        nums[n - 1] = pre[n - 2];
        return nums;
    }
};

File: Leetcode/238/238.go
func productExceptSelf(nums []int) []int {
	preFromLeft := make([]int, len(nums))
	preFromLeft[0] = nums[0]
	for i := 1; i < len(nums); i++ {
		preFromLeft[i] = preFromLeft[i-1] * nums[i]
	}
	preFromRight := make([]int, len(nums))
	preFromRight[len(nums)-1] = nums[len(nums)-1]
	for i := len(nums)-2; i >= 0; i-- {
		preFromRight[i] = preFromRight[i+1] * nums[i]
	}
	answer := make([]int, len(nums))
	answer[0] = preFromRight[1]
	answer[len(nums)-1] = preFromLeft[len(nums)-2]
	for i := 1; i < len(nums)-1; i++ {
		answer[i] = preFromLeft[i-1]*preFromRight[i+1]
	}
	return answer
}